### DISCUSSION TITLE
Make pkg dependent on node-sea

### DISCUSSION THREAD

**Discussion by Sadzurami** [yao-pkg/pkg#97] [Ideas]
Just an idea, that can change and make easier maintaining of this project.

What about significantly change pkg methods for creating executable?
Since node officially supports [sea](https://nodejs.org/api/single-executable-applications.html), pkg can depends of it.

No need to patch node every time. It can reduce required boilerplate for future releases.

- Why not to use node-sea directly?
- Because it requires more additional and not convenient steps to make one single executable, while pkg can make it with one single command.

Key suggestion is - make pkg dependent on node-sea and preserve existing api / cli interfaces.

**Comment by robertsLando** [comment]
@Sadzurami This is the idea for the future, BTW implementing this is still very complicated as sea is still missing support for VFS (virtual file system). Actually it supports adding assets to the file but not multiple js files. In order to fix this there should be a pre-build step using esbuild or bpkg that will take care of bundling everything to a single file but this may require some tweaks in complex applications

**Reply by Sadzurami** [reply]
> @Sadzurami This is the idea for the future, BTW implementing this is still very complicated as sea is still missing support for VFS (virtual file system). Actually it supports adding assets to the file but not multiple js files. In order to fix this there should be a pre-build step using esbuild or bpkg that will take care of bundling everything to a single file but this may require some tweaks in complex applications

yes, exactly what you mention
bundling before creation executable is required step

now, as i know, pkg makes it by itselft
but i don't see any troubles to use es-build, bpkg, or webpack under the hood and have switch to disable this feature (if required)

it all must be easier to maintain, than patching nodejs every release
also, it will bring esm support as well and possibility for other cool features

**Reply by robertsLando** [reply]
@Sadzurami You know that sea doesn't support cross arch/os building as well while pkg does? Also the pre-build step isn't always deterministic, may require some manual tweak expecially when we need to deal with `.node` files. 

For example check: https://github.com/zwave-js/zwave-js-ui/blob/master/esbuild.js#L118

As you can see here creating the single executable file isn't that easy as it require some monkey patch, so I'm not sure how hard it could be to create a single file with a process that handles all the edge cases...

**Reply by Sadzurami** [reply]
> @Sadzurami You know that sea doesn't support cross arch/os building as well while pkg does? Also the pre-build step isn't always deterministic, may require some manual tweak expecially when we need to deal with `.node` files.
> 
> For example check: https://github.com/zwave-js/zwave-js-ui/blob/master/esbuild.js#L118
> 
> As you can see here creating the single executable file isn't that easy as it require some monkey patch, so I'm not sure how hard it could be to create a single file with a process that handles all the edge cases...

Even pkg can't handle all cases with cross arch/os builds. As my experience, i can't build winx64 executable at project with sqlite3, while running that on arm64 mac.

As for tweaks with native modules (.node files) i don't know how it can handle most cases. Maybe you are right and this is deeply, than looks.

**Reply by robertsLando** [reply]
@Sadzurami yeah the entire argument is sooo complicated, trust me. Anyway this is on my TODO list and I may add it with experimental support soon :) 

In order to keep an eye on sea updates check: https://github.com/nodejs/single-executable

**Reply by segevfiner** [reply]
I think it is possible to add additional assets using Node's sea support, meaning it can probably be implemented to use the existing bootstrap entry point of pkg and just having it look for the created VFS assets using the sea API. Instead of just plugging in the user entry point directly as the sea entry point. Might require some work to get this working with code cache and/or snapshot though.

But it will be cool if in the end `pkg` can just use Node's sea support (And maybe bytenode) and work without a patched Node.js executable.

**Reply by robertsLando** [reply]
I added sea support in last 6.0.0 version

**Reply by segevfiner** [reply]
Yeah, but that's without VFS and the whole bootstrap thing pkg does in non-sea mode?

**Reply by robertsLando** [reply]
Yeah as right now there is no real support for it, there is just a support for assets but they still would require user to handle cases where the application is running in a sea bin or plain node as assets have dedicated method to be accessed on sea.

Would be great BTW if this logic could be implemented in NodeJS

**Reply by robertsLando** [reply]
Dealing with some custom implementation of VFS could be a real pain

**Reply by segevfiner** [reply]
You already have one. In the bootstrap.js of pkg. Just a matter of making it get the binary blob it uses from the sea asset instead of the appended blob to the executable I think.

**Reply by robertsLando** [reply]
I would create a dedicated one for sea instead of editing that one as that's huge and actually has lot of useless stuff. 

Do you want to submit a PR?

**Reply by segevfiner** [reply]
Makes sense. But it does have a lot of code for the VFS stuff... So might need to refactor and maybe generate those files using a bundler. (It is also rather messy code IMHO).

I can't guarantee I'll have any time to take a stab at this anytime soon though... I'm overloaded in both personal life and work 😔

**Reply by robertsLando** [reply]
No worries, btw I agree with this, also the bundler could allow us to write them using ts for better type checking
